### ISSUE TITLE
ucx-1.14.0 fails to compile with nvc (NVIDIA HPC SDK C compiler) on aarch64 - apparently incorrect code

### ISSUE THREAD

**Issue by cparrott73** [openucx/ucx#9035]
### Describe the bug
During a recent attempt to build ucx-1.14.0 on aarch64 with nvc 23.3, I got the following error:

```
  CC       bistro/libucm_la-bistro_aarch64.lo
"bistro/bistro_aarch64.c", line 246: error #853: expression must be a pointer to a complete object type
      orig_func    = ucm_bistro_allocate_code(sizeof(*orig_func) + max_code_len +
                                                      ^

1 error detected in the compilation of "bistro/bistro_aarch64.c".
```

### Steps to Reproduce
```
./configure --enable-mt --with-cuda=/proj/cuda/11.0/Linux_aarch64 --without-java --without-go --without-valgrind --prefix=/tmp/ompi4-11 CC=nvc CXX=nvc++
make
```

### Setup and versions
- RHEL 7.6 (Maipo) on aarch64 on Cavium ThunderX2 CPU-based system
- NVIDIA HPC SDK 22.3

### Additional Info

The code fails to compile with nvc at line 246 of the file `src/ucm/bistro/bistro_aarch64.c`:

```
static ucs_status_t
ucm_bistro_relocate_func(const void *func_ptr, size_t patch_len,
                         const char *symbol, void **orig_func_p)
{
    ucm_bistro_aarch64_relocate_context_t aarch64_ctx;
    size_t code_len, prefix_len, max_code_len;
    ucs_status_t status;
    void *orig_func;
    uint8_t regno;

    /* Allocate executable page, calculate the patch size according to worse
       case scenario when need to translate ADRP */
    max_code_len = (patch_len *
                    (sizeof(ucm_bistro_load64_t) / sizeof(ucm_bistro_inst_t))) +
                   sizeof(ucm_bistro_jmp_indirect_t);
    orig_func    = ucm_bistro_allocate_code(sizeof(*orig_func) + max_code_len +
                                            sizeof(ucm_bistro_jmp_indirect_t));
    if (orig_func == NULL) {
        return UCS_ERR_NO_MEMORY;
    }
```

Specifically, this statement:

```
    orig_func    = ucm_bistro_allocate_code(sizeof(*orig_func) + max_code_len +
                                            sizeof(ucm_bistro_jmp_indirect_t));
```

gcc and clang allow this statement, but nvc fails to compile it.  According to our C/C++ language expert, taking sizeof() of a dereferenced void pointer is illegal according to the C standard:

> That is non-standard behavior in both C and C++. In both languages, both (1) dereferencing a pointer-to-void and (2) "sizeof(void)" are errors.

> GCC C: Accepts both (1) and (2) without any messages.
> Clang C: Accepts both (1) and (2) without any messages.
> nvc C: Accepts (2) without a message. Errors on (1) (which is what this bug is about).

Would you please double-check this code for compliance with the C standard, and rework it so that it will compile with nvc?

Thanks in advance.

**Comment by ivankochin** [comment]
@cparrott73

Hi, thank you for posting that!

Could you please share the log of your `configure` script execution?

**Comment by ivankochin** [comment]
We have merged the patch that should solve this: https://github.com/openucx/ucx/pull/9056

Could you please try to reproduce with the latest master.

**Comment by louspe-linaro** [comment]
@ivankochin What release version will that go into? I didn't see it on the release notes for 1.14.1

**Comment by ivankochin** [comment]
> @ivankochin What release version will that go into? I didn't see it on the release notes for 1.14.1

I think it would be included in the 1.16 release only.
